src/meson.build

```diff
@@ -0,0 +1,428 @@
+  endif
+endforeach
+
+if not cdata.has('HAVE_GETOPT_H') or not cc.has_function('getopt_long', prefix: have_prefix, args: cc_defs)
+  src = ['getopt.c', 'getopt1.c']
+  tinc_src += src
+  tincd_src += src
+  sptps_test_src += src
+  sptps_keypair_src += src
+endif
+
+opt_miniupnpc = get_option('miniupnpc')
+if not opt_miniupnpc.disabled()
+  dep_miniupnpc = dependency('miniupnpc', required: false, static: static)
+  if not dep_miniupnpc.found()
+    # No pkg-config files on MinGW
+    dep_miniupnpc = cc.find_library('miniupnpc', required: opt_miniupnpc, static: static)
+  endif
+  if dep_miniupnpc.found()
+    tincd_src += 'upnp.c'
+    tincd_dep += [
+      dependency('threads', static: static),
+      dep_miniupnpc,
+    ]
+    if static
+      tincd_cc_flags += '-DMINIUPNP_STATICLIB'
+    endif
+    cdata.set('HAVE_MINIUPNPC', 1)
+  endif
+endif
+
+opt_curses = get_option('curses')
+if opt_curses.auto() and os_name == 'windows'
+  message('curses does not link under MinGW')
+else
+  # The meta-dependency covers more alternatives, but is only available in 0.54+
+  curses_name = meson.version().version_compare('>=0.54') ? 'curses' : 'ncurses'
+  dep_curses = dependency(curses_name, required: opt_curses, static: static)
+  if dep_curses.found()
+    cdata.set('HAVE_CURSES', 1)
+    tinc_dep += dep_curses
+  endif
+endif
+
+# Some distributions do not supply pkg-config files for readline
+opt_readline = get_option('readline')
+if opt_readline.auto() and os_name == 'windows'
+  message('readline does not link under MinGW')
+else
+  dep_readline = dependency('readline', required: opt_readline, static: static)
+  if not dep_readline.found()
+    dep_readline = cc.find_library('readline', required: opt_readline, static: static)
+  endif
+  if dep_readline.found() and cc.has_header('readline/readline.h', dependencies: dep_readline) and cc.has_header('readline/history.h', dependencies: dep_readline)
+    cdata.set('HAVE_READLINE', 1)
+    tinc_dep += dep_readline
+  endif
+endif
+
+dep_zlib = dependency('zlib', required: get_option('zlib'), static: static)
+if dep_zlib.found()
+  cdata.set('HAVE_ZLIB', 1)
+  tincd_dep += dep_zlib
+endif
+
+opt_lzo = get_option('lzo')
+if not opt_lzo.disabled()
+  dep_lzo = dependency('lzo2', required: false, static: static)
+  if not dep_lzo.found()
+    dep_lzo = cc.find_library('lzo2', required: opt_lzo, static: static)
+  endif
+  if dep_lzo.found() and cc.has_function('lzo1x_1_compress', dependencies: dep_lzo) and cc.has_header('lzo1x.h', dependencies: dep_lzo)
+    cdata.set('HAVE_LZO', 1)
+    tincd_dep += dep_lzo
+  endif
+endif
+
+opt_lz4 = get_option('lz4')
+if not opt_lz4.disabled()
+  dep_lz4 = dependency('liblz4', required: false, static: static)
+  if dep_lz4.found()
+    tincd_dep += dep_lz4
+    cdata.set('HAVE_LZ4', 1)
+    cdata.set('LZ4_H', '<lz4.h>')
+  elif not opt_lz4.auto()
+    tincd_src += 'lib/lz4/lz4.c'
+    cdata.set('HAVE_LZ4', 1)
+    cdata.set('LZ4_H', '"lib/lz4/lz4.h"')
+  endif
+endif
```

Comment:
Upstream lz4 has a meson.build, which is wrapped in the Meson WrapDB.

```
mkdir -p subprojects
meson wrap install lz4
```

And then `dependency('lz4', required: get_option('lz4'))` can automatically fall back to the subproject, download it on demand (if wrap-mode doesn't specify `nodownload`/`nofallback`) and build it as a private copy.

Reply:
Great, happy to help.

There are other wraps in the WrapDB which you may want to take advantage of too. :)